// @onecli-sh/sdk/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AgentsClient: () => AgentsClient,
  ContainerClient: () => ContainerClient,
  OneCLI: () => OneCLI,
  OneCLIError: () => OneCLIError,
  OneCLIRequestError: () => OneCLIRequestError
});
module.exports = __toCommonJS(index_exports);

// src/errors.ts
var OneCLIError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "OneCLIError";
  }
};
var OneCLIRequestError = class extends Error {
  url;
  statusCode;
  constructor(message, requestData) {
    super(
      `[URL=${requestData.url}] [StatusCode=${requestData.statusCode}] ${message}`
    );
    this.name = "OneCLIRequestError";
    this.url = requestData.url;
    this.statusCode = requestData.statusCode;
  }
};
function toOneCLIError(error) {
  if (error instanceof OneCLIError || error instanceof OneCLIRequestError) {
    return error;
  }
  if (error instanceof Error) {
    return new OneCLIError(error.message);
  }
  return new OneCLIError(String(error));
}

// src/container/ca.ts
var import_fs = require("fs");
var import_os = require("os");
var import_path = require("path");
var SYSTEM_CA_PATHS = [
  "/etc/ssl/cert.pem",
  // macOS
  "/etc/ssl/certs/ca-certificates.crt",
  // Debian / Ubuntu
  "/etc/pki/tls/certs/ca-bundle.crt"
  // RHEL / CentOS / Fedora
];
function writeCaCertificate(caCertificate) {
  const outPath = (0, import_path.join)((0, import_os.tmpdir)(), "onecli-proxy-ca.pem");
  (0, import_fs.writeFileSync)(outPath, caCertificate);
  return outPath;
}
function buildCombinedCaBundle(caCertificate) {
  for (const sysPath of SYSTEM_CA_PATHS) {
    try {
      const sysCa = (0, import_fs.readFileSync)(sysPath, "utf8");
      const combined = sysCa.trimEnd() + "\n" + caCertificate.trimEnd() + "\n";
      const outPath = (0, import_path.join)((0, import_os.tmpdir)(), "onecli-combined-ca.pem");
      (0, import_fs.writeFileSync)(outPath, combined);
      return outPath;
    } catch {
      continue;
    }
  }
  return null;
}

// src/container/index.ts
var ContainerClient = class {
  baseUrl;
  apiKey;
  timeout;
  constructor(baseUrl, apiKey, timeout) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.apiKey = apiKey;
    this.timeout = timeout;
  }
  /**
   * Fetch the raw container configuration from OneCLI.
   */
  getContainerConfig = async (agent) => {
    const url = agent ? `${this.baseUrl}/api/container-config?agent=${encodeURIComponent(agent)}` : `${this.baseUrl}/api/container-config`;
    try {
      const headers = {};
      if (this.apiKey) {
        headers["Authorization"] = `Bearer ${this.apiKey}`;
      }
      const res = await fetch(url, {
        headers,
        signal: AbortSignal.timeout(this.timeout)
      });
      if (!res.ok) {
        throw new OneCLIRequestError(
          `OneCLI returned ${res.status} ${res.statusText}`,
          { url, statusCode: res.status }
        );
      }
      return await res.json();
    } catch (error) {
      if (error instanceof OneCLIError || error instanceof OneCLIRequestError) {
        throw error;
      }
      throw toOneCLIError(error);
    }
  };
  /**
   * Fetch the container config from OneCLI and push the corresponding
   * `-e` and `-v` flags onto the Docker `run` argument array.
   *
   * Returns `true` if OneCLI was reachable and config was applied,
   * `false` otherwise (graceful degradation).
   */
  applyContainerConfig = async (args, options) => {
    const { combineCaBundle = true, addHostMapping = true, agent } = options ?? {};
    let config;
    try {
      config = await this.getContainerConfig(agent);
    } catch {
      return false;
    }
    for (const [key, value] of Object.entries(config.env)) {
      args.push("-e", `${key}=${value}`);
    }
    const hostCaPath = writeCaCertificate(config.caCertificate);
    args.push(
      "-v",
      `${hostCaPath}:${config.caCertificateContainerPath}:ro`
    );
    if (combineCaBundle) {
      const combinedPath = buildCombinedCaBundle(config.caCertificate);
      if (combinedPath) {
        args.push("-e", "SSL_CERT_FILE=/tmp/onecli-combined-ca.pem");
        args.push("-v", `${combinedPath}:/tmp/onecli-combined-ca.pem:ro`);
      }
    }
    if (addHostMapping && process.platform === "linux") {
      args.push("--add-host", "host.docker.internal:host-gateway");
    }
    return true;
  };
};

// src/agents/index.ts
var AgentsClient = class {
  baseUrl;
  apiKey;
  timeout;
  constructor(baseUrl, apiKey, timeout) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.apiKey = apiKey;
    this.timeout = timeout;
  }
  /**
   * Create a new agent.
   */
  createAgent = async (input) => {
    const url = `${this.baseUrl}/api/agents`;
    const headers = {
      "Content-Type": "application/json"
    };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    try {
      const res = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify(input),
        signal: AbortSignal.timeout(this.timeout)
      });
      if (!res.ok) {
        throw new OneCLIRequestError(
          `OneCLI returned ${res.status} ${res.statusText}`,
          { url, statusCode: res.status }
        );
      }
      return await res.json();
    } catch (error) {
      if (error instanceof OneCLIError || error instanceof OneCLIRequestError) {
        throw error;
      }
      throw toOneCLIError(error);
    }
  };
  /**
   * Ensure an agent exists. Creates it if missing, returns normally if it already exists.
   * Unlike `createAgent`, this method treats a 409 conflict as success.
   */
  ensureAgent = async (input) => {
    try {
      await this.createAgent(input);
      return { name: input.name, identifier: input.identifier, created: true };
    } catch (error) {
      if (error instanceof OneCLIRequestError && error.statusCode === 409) {
        return {
          name: input.name,
          identifier: input.identifier,
          created: false
        };
      }
      throw error;
    }
  };
};

// src/client.ts
var DEFAULT_URL = "https://app.onecli.sh";
var DEFAULT_TIMEOUT = 5e3;
var OneCLI = class {
  containerClient;
  agentsClient;
  constructor(options = {}) {
    const apiKey = options.apiKey ?? process.env.ONECLI_API_KEY ?? "";
    const url = options.url ?? process.env.ONECLI_URL ?? DEFAULT_URL;
    const timeout = options.timeout ?? DEFAULT_TIMEOUT;
    this.containerClient = new ContainerClient(url, apiKey, timeout);
    this.agentsClient = new AgentsClient(url, apiKey, timeout);
  }
  /**
   * Fetch the raw container configuration from OneCLI.
   */
  getContainerConfig = (agent) => {
    return this.containerClient.getContainerConfig(agent);
  };
  /**
   * Fetch config and apply `-e` / `-v` flags to a Docker `run` argument array.
   * Returns `true` on success, `false` on failure (graceful degradation).
   */
  applyContainerConfig = (args, options) => {
    return this.containerClient.applyContainerConfig(args, options);
  };
  /**
   * Create a new agent.
   */
  createAgent = (input) => {
    return this.agentsClient.createAgent(input);
  };
  /**
   * Ensure an agent exists. Creates it if missing, returns normally if it already exists.
   */
  ensureAgent = (input) => {
    return this.agentsClient.ensureAgent(input);
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentsClient,
  ContainerClient,
  OneCLI,
  OneCLIError,
  OneCLIRequestError
});
